Remove public constructor requirement

justinjstark opened the pull request:
Instead, use cached delegate for constructor call.

justinjstark committed:
Remove public constructor requirement

Instead, use cached delegate for constructor call.

justinjstark commented:
I updated pull request #3 to use Expression.New and to cache the delegate for the constructor call to a parameterless constructor.

justinjstark commented:
Solves Issue #1

mattnischan commented on Immutable.Net/Immutable.Net/ImmutableBuilder`T.cs:
```diff
@@ -35,7 +35,7 @@ public static ImmutableBuilder<T> Create(T self)
         public ImmutableBuilder() 
         {
-            this.self = new T();
+            self = Expression.Lambda<Func<T>>(Expression.New(typeof(T))).Compile().Invoke();
```
This needs to be cached and then reused.

mattnischan commented on Immutable.Net/Immutable.Net/Immutable`T.cs:
```diff
@@ -79,7 +86,7 @@ private Immutable(T self)
         private Immutable(SerializationInfo info, StreamingContext context)
         {
-            self = new T();
+            self = createDelegate.Invoke();
```
If this is the first constructor called, createDelegate will be null.

mattnischan commented on Immutable.Net/Immutable.Net/Immutable`T.cs:
```diff
@@ -59,7 +64,9 @@ private static class Accessor<TValue>
         public Immutable() 
         {
-            self = new T();
+            createDelegate = Expression.Lambda<Func<T>>(Expression.New(typeof(T))).Compile();
```
If the createDelegate already exists, we don't need to recreate it here. We should reuse this as well.

mattnischan commented:
Definitely on the right track. Just a few small issues and then this can be merged.

You can make the changes to this PR, no need to open another.

justinjstark commented:
I fixed the issues you mentioned and also added a couple tests to verify that an `Immutable` and `ImmutableBuilder` can be instantiated using a class with a private constructor. I also pushed the `creationDelegate` creation back to the `DelegateBuilder` so both the `Immutable` and `ImmutableBuilder` can use the common builder.

I didn't realize I could update a PR. I'm pretty new to PRs on github. I squashed the latest update with the previous commit. I'm not sure what the etiquette is regarding that.

Let me know if there is anything else I should fix/change. Thank you for your patience and guidance on this.

mattnischan commented on Immutable.Net/Immutable.Net/ImmutableBuilder`T.cs:
```diff
@@ -35,7 +40,12 @@ public static ImmutableBuilder<T> Create(T self)
-            this.self = new T();
+            if (creationDelegate == null)
+            {
+                creationDelegate = Expression.Lambda<Func<T>>(Expression.New(typeof(T))).Compile();
```
You can use your `DelegateBuilder` method here too.

mattnischan commented:
Of course! Yeah, generally most people squash their commits for PRs into a single commit. Thanks for contributing!

justinjstark commented:
Fixed per your comment.

mattnischan commented:
Looks good here.
